pkg/utils/permission.go

```diff
@@ -0,0 +1,275 @@
+	WithStaticRoleBinding(clusterRole *rbacv1.RoleBinding) RBACPermissionBuilder
+
+	// Build wraps up the builder chain, and return a agent.PermissionConfigFunc.
+	Build() agent.PermissionConfigFunc
+}
+
+var _ RBACPermissionBuilder = &permissionBuilder{}
+
+type permissionBuilder struct {
+	kubeClient kubernetes.Interface
+	u          *unionPermissionBuilder
+}
+
+// NewRBACPermissionConfigBuilder instantiates a default RBACPermissionBuilder.
+func NewRBACPermissionConfigBuilder(kubeClient kubernetes.Interface) RBACPermissionBuilder {
+	return &permissionBuilder{
+		u:          &unionPermissionBuilder{},
+		kubeClient: kubeClient,
+	}
+}
+
+func (p *permissionBuilder) BindClusterRoleToUser(clusterRole *rbacv1.ClusterRole, username string) RBACPermissionBuilder {
+	return p.WithStaticClusterRole(clusterRole).
+		WithStaticClusterRoleBinding(&rbacv1.ClusterRoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: clusterRole.Name, // The same name as the cluster role
+			},
+			RoleRef: rbacv1.RoleRef{
+				Kind: "ClusterRole",
+				Name: clusterRole.Name,
+			},
+			Subjects: []rbacv1.Subject{
+				{
+					Kind: rbacv1.UserKind,
+					Name: username,
+				},
+			},
+		})
+}
+
+func (p *permissionBuilder) BindClusterRoleToGroup(clusterRole *rbacv1.ClusterRole, userGroup string) RBACPermissionBuilder {
+	return p.WithStaticClusterRole(clusterRole).
+		WithStaticClusterRoleBinding(&rbacv1.ClusterRoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: clusterRole.Name, // The same name as the cluster role
+			},
+			RoleRef: rbacv1.RoleRef{
+				Kind: "ClusterRole",
+				Name: clusterRole.Name,
+			},
+			Subjects: []rbacv1.Subject{
+				{
+					Kind: rbacv1.GroupKind,
+					Name: userGroup,
+				},
+			},
+		})
+}
+
+func (p *permissionBuilder) BindRoleToUser(role *rbacv1.Role, username string) RBACPermissionBuilder {
+	return p.WithStaticRole(role).
+		WithStaticRoleBinding(&rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: role.Name, // The same name as the cluster role
+			},
+			RoleRef: rbacv1.RoleRef{
+				Kind: "Role",
+				Name: role.Name,
+			},
+			Subjects: []rbacv1.Subject{
+				{
+					Kind: rbacv1.UserKind,
+					Name: username,
+				},
+			},
+		})
+}
+
+func (p *permissionBuilder) BindRoleToGroup(role *rbacv1.Role, userGroup string) RBACPermissionBuilder {
+	return p.WithStaticRole(role).
+		WithStaticRoleBinding(&rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: role.Name, // The same name as the cluster role
+			},
+			RoleRef: rbacv1.RoleRef{
+				Kind: "Role",
+				Name: role.Name,
+			},
+			Subjects: []rbacv1.Subject{
+				{
+					Kind: rbacv1.GroupKind,
+					Name: userGroup,
+				},
+			},
+		})
+}
+
+func (p *permissionBuilder) WithStaticClusterRole(clusterRole *rbacv1.ClusterRole) RBACPermissionBuilder {
+	p.u.fns = append(p.u.fns, func(cluster *clusterv1.ManagedCluster, addon *addonapiv1alpha1.ManagedClusterAddOn) error {
+		ensureAddonOwnerReference(&clusterRole.ObjectMeta, addon)
```

Comment:
We cannot add this since this is a clusterrole

Reply:
removed ownerref